Test/NumberTheory/FastPower.test.cpp

```diff
@@ -0,0 +1,28 @@
+#define FAST_POWER_TEST
+#define CATCH_CONFIG_MAIN
+
+#include "../catch.hpp"
+#include "../../NumberTheory/FastPower.cpp"
+
+
+TEST_CASE("Base cases", "[fast_power]") {
+    //REQUIRE(fast_power(0,0) == undefined);
+    REQUIRE(fast_power(0,1) == 0);
+    REQUIRE(fast_power(1,0) == 1);
+    REQUIRE(fast_power(1,1) == 1);
+}
+
+TEST_CASE("Normal cases", "[fast_power]") {
+    REQUIRE(fast_power(2,2) == 4);
+    REQUIRE(fast_power(2,4) == 16);
+    REQUIRE(fast_power(3,4) == 81);
+    REQUIRE(fast_power(7,9) == 40353607);
+    REQUIRE(fast_power(15,10) == 576650390625);
+}
+
+TEST_CASE("Overflow cases", "[fast_power]") {
+    REQUIRE(fast_power(2,100) == 68719476736);
+    REQUIRE(fast_power(10,99) == 4440381706574496940U);
```

Comment:
Yes, passing a third parameter is the right way instead of having 2 similar functions. But use `ULLONG_MAX` as the mod value in the `else` case instead of 10<sup>18</sup>.